src/Lykke.AlgoStore.CSharp.AlgoTemplate.Services/Utils/CandleHistoryBatchEnumerator.cs

```diff
@@ -108,24 +110,46 @@ private void IncrementBuffer()
                     _isLastBuffer = true;
                 }
 
-                FillBuffer();
+                await FillBuffer();
             }
             while ((_buffer == null || _buffer.Count == 0) && !_isLastBuffer);
         }
 
         /// <summary>
         /// Fills the buffer from the history service
         /// </summary>
-        private void FillBuffer()
+        private async Task FillBuffer()
         {
-            var task = _historyClient.GetCandles(_currentTimestamp, _nextTimestamp,
-                                                 _candlesHistoryRequest.IndicatorName,
-                                                 _candlesHistoryRequest.AuthToken);
+            IEnumerable<Service.History.Client.Models.Candle> history = null;
 
-            var history = task.ConfigureAwait(false).GetAwaiter().GetResult();
+            while (history == null)
```

Comment:
Why check in a while loop? if you expect history == null in a while loop, that suggests that it may remain null for a few loop iterations, which would mean you would eventually get null reference exception on .ToList()

Reply:
history will be null when .GetCandles throws an exception. In all other cases it will have a value and ToList will work properly. This while loop is here to have the operation retry until it returns a proper value.

Reply:
ок.